src/hooks/useTokenInfo.ts

```diff
@@ -0,0 +1,145 @@
+
+                    if (!symbol) {
+                        throw new Error('Not an ERC-20 token')
+                    }
+
+                    const name = await contract.name()
+
+                    const decimals = await contract.decimals()
+
+                    resolve({
+                        decimals,
+                        name,
+                        symbol,
+                    })
+                } catch (e) {
+                    reject(e)
+                }
+            })
+        })
+
+        tokenInfoCache[key] = promise
+
+        setTimeout(async () => {
+            let next: TokenInfo | null = null
+
+            try {
+                next = await promise
+            } catch (e) {
+                /**
+                 * If we get a rejection from the above promise we simply
+                 * cache `null`. Not the bestest of ways, but we can
+                 * improve as we go forward.
+                 */
+            }
+
+            tokenInfoCache[key] = next
+        })
+
+        return promise
+    }
+
+    return current
+}
+
+/**
+ * A hook that fetches and returns a `TokenInfo` object for a given pair of token address and chain id.
+ * @param tokenAddress Address of the ERC-20 token contract.
+ * @param chainId Chain/network id.
+ * @returns nullish value or a `TokenInfo` object.
+ * - `null` if the info has been fetched and the result indicated it's not an ERC-20 token,
+ * - `undefined` if the info hasn't been fetched yet (or is being fetched), and
+ * - `TokenInfo` object otherwise.
+ */
+export default function useTokenInfo(
+    tokenAddress: string | undefined,
+    chainId: number | undefined,
+) {
+    const key =
+        tokenAddress && typeof chainId !== 'undefined'
+            ? entryKey(tokenAddress, chainId)
+            : null
+
+    const [tokenInfo, setTokenInfo] = useState<TokenInfo | null | undefined>(
+        key
+            ? tokenInfoCache[key] instanceof Promise
+                ? undefined
+                : (tokenInfoCache[key] as TokenInfo | null | undefined)
+            : null,
+    )
+
+    /**
+     * Undefined token info causes a fetch *attempt*. Since we cache both the values
+     * and their pending promises there's no redundant loading.
+     */
+    const shouldFetch = typeof tokenInfo === 'undefined'
+
+    useEffect(() => {
+        let mounted = true
+
+        let result: TokenInfo | null = null
+
+        if (shouldFetch) {
+            setTimeout(async () => {
+                try {
+                    result = await getTokenInfo(tokenAddress, chainId)
+                } catch (e) {
+                    console.warn(
+                        `Failed to load token information for ${tokenAddress}`,
+                        e,
+                    )
+                }
+
+                if (mounted) {
+                    setTokenInfo(result)
+                }
+            })
+        }
+
+        return () => {
+            mounted = false
```

Comment:
useIsMounted hook does that for you

Reply:
No, it doesn't.

```ts
function useFoo(tokenAddress: string) {
    const isMounted = useIsMounted()

    const [foo, setFoo] = useState<number>()

    useEffect(() => {
        setTimeout(async () => {
            const newFoo = Math.random()

            await new Promise((resolve) => void setTimeout(resolve))

            if (isMounted()) {
                /**
                 * isMounted() is true for all different `tokenAddress` values, and multiple
                 * async workflows can do work at the same time! Scenario:
                 *
                 * 1. Render `useFoo` with tokenAddress "A".
                 * 2. Rerender (w/o unmounting) the same "instance" of it with tokenAddress "B".
                 *
                 * If for whatever reason fetching for "A" finishes AFTER fetching for "B"
                 * then your UI will display values that don't match the input (FIFO).
                 */
                setFoo(newState)
            }
        })
    }, [tokenAddress, isMounted])

    return foo
}
```

`useIsMounted` has to go. It can cause a wide range of WTFs.

Reply:
I just tested this useFoo hook and it works perfectly with useIsMounted().

"isMounted() is true for all different `tokenAddress` values" <-- if you use setTimeout with undefined delay time then yeah - it will always return true, cause the hook will be still mounted. I tried with a delay of 5 seconds so in the meantime I could leave the page that uses the useFoo hook, so it gets dismounted and then isMounted() returns false. Really - I don't see any issue with useIsMounted() here

Reply:
We're talking about different things.

Of course if you unmount it'll switch to false. Happy path is fine. I'm talking rerendering for different props.

Check out my playground
https://codesandbox.io/s/useismounted-playground-qxvv3f?file=/src/App.tsx

Reply:
Ok now I see it! Man... React sucks :)) Angular is built on top of RxJS which does for instance handle cancelling of asynchronous data streams for you :)
BTW - good idea with replacing the implementation of useIsMounted!